crates/notion-core/src/fs.rs

```diff
@@ -18,36 +16,20 @@ pub fn touch(path: &Path) -> Fallible<File> {
     File::open(path).unknown()
 }
 
-#[derive(Debug, Fail, NotionFail)]
-#[fail(display = "Could not create directory {}: {}", dir, error)]
-#[notion_fail(code = "FileSystemError")]
-pub(crate) struct CreateDirError {
-    pub(crate) dir: String,
-    pub(crate) error: String,
-}
-
-impl CreateDirError {
-    pub(crate) fn for_dir(dir: String) -> impl FnOnce(&io::Error) -> CreateDirError {
-        move |error| CreateDirError {
-            dir,
-            error: error.to_string(),
-        }
+fn error_for_dir(dir: String) -> impl FnOnce(&io::Error) -> ErrorDetails {
+    move |error| ErrorDetails::CreateDirError {
+        dir,
+        error: error.to_string(),
```

Comment:
This is a great idea for the general case! I think I'll update the plan for adapting the `.unknown()` calls to use an `impl From` approach, as it makes the call to `with_context` more expressive (`.with_context(ErrorDetails::from)`).

In this case, we actually need a bespoke function, because we need the additional context provided by the `dir` parameter to create the error.

I agree about losing the detail of the underlying error. There are a number of the existing `NotionFail`s that incorporate an `error` string, rather than the underlying error. Cleaning those up to more accurately include the full error in the failure is a good next step.

Reply:
Thinking on this more, we need to make sure we align what the errors are supposed to represent and don't throw away context. Specifically, the underlying errors that we are getting are almost always descriptions of what went wrong, while the `NotionError` we want to present to the user is "This is what we were trying to do, and here's what happened when we did that". We're always going to be adding that additional context about what we were trying to do, so I'm not sure that there will be a good general conversion between underlying errors and `NotionError`s.

That said, it's definitely good to keep an eye out for spots where we could use an idiomatic construct like `impl From`, as opposed to rolling our own.

Reply:
Yeah, I figured there'd be limitations to what we can do—and you certainly have a better sense of that space than I do at this point! Even if we end up throwing away most of it, using richer data to construct our own error when it's available (i.e. *not* just doing a `to_string()` if there's more we can use) seems like it'll be a nice win.